tasks/codecov.py

```diff
@@ -89,14 +97,115 @@ def __exit__(self, *_):
 
 @task
 def codecov(
-    ctx,
+    ctx: Context,
+    pull_coverage_cache: bool = False,
+    push_coverage_cache: bool = False,
+    debug: bool = False,
 ):
     """
-    Uploads coverage data of all modules.
+    Uploads coverage data of all modules to Codecov.
     This expects that the coverage files have already been generated by
     inv test --coverage.
+
+    Flags:   --pull-coverage-cache: [For dev branches] Pull the coverage cache from main parent commit.
+             --push-coverage-cache: [For main]         Push the coverage cache to the S3 bucket.
+             --debug:                                  Print debug information.
     """
+    if pull_coverage_cache and push_coverage_cache:
+        raise Exit(
+            color_message("Error: Can't use both --pull-missing-coverage and --push-coverage-cache flags.", Color.RED),
+            code=1,
+        )
     distro_tag = get_distro()
     codecov_binary = "codecov" if platform.system() != "Windows" else "codecov.exe"
+
     with gitlab_section("Upload coverage reports to Codecov", collapsed=True):
+        if pull_coverage_cache:
+            apply_missing_coverage(ctx, from_commit_sha=get_main_parent_commit(ctx), debug=debug)
+        if push_coverage_cache:
+            upload_coverage_to_s3(ctx)
         ctx.run(f"{codecov_binary} -f {PROFILE_COV} -F {distro_tag}", warn=True)
+
+
+def produce_coverage_tar(files, archive_name):
+    """
+    Produce a tgz file containing all coverage files.
+    """
+    with tarfile.open(archive_name, "w:gz") as tgz:
+        for f in files:
+            tgz.add(f)
+    print(color_message(f"Successfully created {archive_name}", Color.GREEN))
+
+
+def _get_coverage_cache_uri():
+    if BUCKET_CI_VAR not in os.environ:
+        raise Exit(color_message(f"Error: the {BUCKET_CI_VAR} environment variable is not set.", Color.RED), code=1)
+    return f"{os.environ[BUCKET_CI_VAR]}/coverage-cache"
+
+
+@task
+def upload_coverage_to_s3(ctx: Context):
+    """
+    Create an archive with all the coverage.out files from the inv test --coverage command.
+    Then upload the archive to the dd-ci-persistent-artefacts-build-stable S3 bucket.
+    """
+
+    # Find all coverage files in the project and put them in a tgz archive
+    cov_files = sorted(pathlib.Path(".").rglob(PROFILE_COV))
+    produce_coverage_tar(cov_files, COV_ARCHIVE_NAME)
+
+    # Upload the archive to S3
+    cache_uri = _get_coverage_cache_uri()
+    commit_sha = os.getenv("CI_COMMIT_SHA") or get_commit_sha(ctx)
+    if ctx.run(f"{AWS_CMD} s3 cp {COV_ARCHIVE_NAME} {cache_uri}/{commit_sha}/", echo=True, warn=True):
+        print(
+            color_message(
+                f'Successfully uploaded coverage cache to {cache_uri}/{commit_sha}/{COV_ARCHIVE_NAME}', Color.GREEN
+            )
+        )
+    else:
+        raise Exit(
+            color_message(f"Failed to upload coverage cache to {cache_uri}/{commit_sha}/{COV_ARCHIVE_NAME}", Color.RED),
```

Comment:
What do you think about printing the error returned by the command. I feel like it would be useful to understand what is happening in case of error

Reply:
The error is already being printed, thanks to the `warn=True` of `ctx.run`, [example here](https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/547852697)